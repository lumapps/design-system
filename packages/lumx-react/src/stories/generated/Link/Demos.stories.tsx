/**
 * File generated when storybook is started. Do not edit directly!
 */
export default { title: 'LumX components/link/Demos' };

export { App as Default } from './default';
export { App as Implicit } from './implicit';
export { App as Sizes } from './sizes';
