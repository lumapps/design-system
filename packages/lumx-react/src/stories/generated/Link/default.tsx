../../../../../site-demo/content/product/components/link/react/default.tsx