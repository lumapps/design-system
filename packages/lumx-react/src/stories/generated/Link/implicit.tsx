../../../../../site-demo/content/product/components/link/react/implicit.tsx