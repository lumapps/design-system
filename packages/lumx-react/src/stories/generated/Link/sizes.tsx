/home/gcornut/git/dup-design-system/packages/site-demo/content/product/components/link/react/sizes.tsx