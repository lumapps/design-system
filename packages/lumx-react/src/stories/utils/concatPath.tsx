export type PathPart = string | number;
export type Path = PathPart | Array<PathPart>;

/**
 * Concat flatten object path
 *
 * @example concatPath('foo', 'bar') // => 'foo.bar'
 * @example concatPath(['foo', 0]) // => 'foo[0]'
 * @example concatPath('foo', 0, ['bar']) // => 'foo[0].bar'
 */
export const concatPath = (...prefix: Path[]) => {
    const [first, ...rest] = prefix.flat();
    return rest.reduce<string>((acc, part) => {
        if (typeof part === 'number') return `${acc}[${part}]`;
        return `${acc}.${part}`;
    }, String(first));
};
