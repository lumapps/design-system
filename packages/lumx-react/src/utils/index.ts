/**
 * WARNING: All modules exported here are exposed to NPM in '@lumx/react/utils'.
 */

export { ClickAwayProvider } from './ClickAwayProvider';
export { Portal, type PortalProps, type PortalInit, PortalProvider, type PortalProviderProps } from './Portal';
