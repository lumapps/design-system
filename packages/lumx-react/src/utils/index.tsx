export * from '@lumx/core/js/utils';

export * from './getRootClassName';

export * from './htmlDecode';

export * from './partitionMulti';

export * from './type';
