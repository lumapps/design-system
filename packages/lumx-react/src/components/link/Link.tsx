import React, { forwardRef, RefObject, useMemo } from 'react';

import isEmpty from 'lodash/isEmpty';

import classNames from 'classnames';

import { Color, ColorVariant, Icon, Size, Typography } from '@lumx/react';
import { Comp, GenericProps, getRootClassName, handleBasicClasses } from '@lumx/react/utils';
import { renderLink } from '@lumx/react/utils/renderLink';

type HTMLAnchorProps = React.DetailedHTMLProps<React.AnchorHTMLAttributes<HTMLAnchorElement>, HTMLAnchorElement>;

/**
 * Defines the props of the component.
 */
export interface LinkProps extends GenericProps {
    /** Color variant. */
    color?: Color;
    /** Lightened or darkened variant of the selected icon color. */
    colorVariant?: ColorVariant;
    /** Link href. */
    href?: HTMLAnchorProps['href'];
    /** Whether the component is disabled or not. */
    isDisabled?: boolean;
    /** Left icon (SVG path). */
    leftIcon?: string;
    /** Custom react component for the link (can be used to inject react router Link). */
    linkAs?: 'a' | any;
    /** Right icon (SVG path). */
    rightIcon?: string;
    /** Link target. */
    target?: HTMLAnchorProps['target'];
    /** Typography variant. */
    typography?: Typography;
}

/**
 * Component display name.
 */
const COMPONENT_NAME = 'Link';

/**
 * Component default class name and class prefix.
 */
const CLASSNAME = getRootClassName(COMPONENT_NAME);

const getIconSize = (typography?: Typography) => {
    switch (typography) {
        case Typography.display1:
            return Size.m;

        case Typography.headline:
        case Typography.title:
        case Typography.body2:
        case Typography.subtitle2:
            return Size.s;

        case Typography.body1:
        case Typography.subtitle1:
            return Size.xs;

        case Typography.caption:
        case Typography.overline:
            return Size.xxs;

        default:
            return Size.s;
    }
};

/**
 * Link component.
 *
 * @param  props Component props.
 * @param  ref   Component ref.
 * @return React element.
 */
export const Link: Comp<LinkProps, HTMLAnchorElement | HTMLButtonElement> = forwardRef((props, ref) => {
    const {
        children,
        className,
        color,
        colorVariant,
        disabled,
        isDisabled = disabled,
        href,
        leftIcon,
        linkAs,
        rightIcon,
        target,
        typography,
        ...forwardedProps
    } = props;
    const renderedChildren = useMemo(
        () => (
            <>
                {leftIcon && !isEmpty(leftIcon) && (
                    <Icon icon={leftIcon} className={`${CLASSNAME}__left-icon`} size={getIconSize(typography)} />
                )}

                {children && (
                    <span
                        className={classNames(`${CLASSNAME}__content`, {
                            [`lumx-typography-${typography}`]: typography,
                        })}
                    >
                        {children}
                    </span>
                )}

                {rightIcon && !isEmpty(rightIcon) && (
                    <Icon icon={rightIcon} className={`${CLASSNAME}__right-icon`} size={getIconSize(typography)} />
                )}
            </>
        ),
        [leftIcon, typography, children, rightIcon],
    );

    /**
     * If there is no linkAs prop and no href, we returned a <button> instead of a <a>.
     * If the component is disabled, we also returned a <button> since disabled is not compatible with <a>.
     */
    if ((!linkAs && isEmpty(href)) || isDisabled) {
        return (
            <button
                type="button"
                {...forwardedProps}
                ref={ref as RefObject<HTMLButtonElement>}
                disabled={isDisabled}
                className={classNames(className, handleBasicClasses({ prefix: CLASSNAME, color, colorVariant }))}
            >
                {renderedChildren}
            </button>
        );
    }
    return renderLink(
        {
            linkAs,
            ...forwardedProps,
            href,
            target,
            className: classNames(className, handleBasicClasses({ prefix: CLASSNAME, color, colorVariant })),
            ref: ref as RefObject<HTMLAnchorElement>,
        },
        renderedChildren,
    );
});
Link.displayName = COMPONENT_NAME;
Link.className = CLASSNAME;
