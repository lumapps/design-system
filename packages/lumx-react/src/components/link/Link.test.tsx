import React, { ReactElement } from 'react';

import { mount, shallow } from 'enzyme';
import 'jest-enzyme';

import { ColorPalette, ColorVariant, Typography } from '@lumx/react';
import { commonTestsSuite, Wrapper } from '@lumx/react/testing/utils';
import { Link, LinkProps } from './Link';

const CLASSNAME = Link.className as string;

type SetupProps = Partial<LinkProps>;

/**
 * Mounts the component and returns common DOM elements / data needed in multiple tests further down.
 */
const setup = ({ ...propsOverride }: SetupProps = {}, shallowRendering = true) => {
    const props: LinkProps = {
        ...propsOverride,
    };
    const renderer: (el: ReactElement) => Wrapper = shallowRendering ? shallow : mount;
    const wrapper: Wrapper = renderer(<Link {...props} />);

    return { props, wrapper };
};

describe(`<${Link.displayName}>`, () => {
    describe('Snapshots and structure', () => {
        it('should render correctly', () => {
            const { wrapper } = setup({ href: 'https://google.com' });
            expect(wrapper).toMatchSnapshot();
        });

        it('should render color & color variant', () => {
            const { wrapper } = setup({
                href: 'https://google.com',
                color: ColorPalette.primary,
                colorVariant: ColorVariant.D1,
            });
            expect(wrapper).toMatchSnapshot();
        });

        it('should render typography', () => {
            const { wrapper } = setup({ href: 'https://google.com', typography: Typography.title });
            expect(wrapper).toMatchSnapshot();
        });

        it('should render a button if href attribute is missing', () => {
            const { wrapper } = setup();
            expect(wrapper).toMatchSnapshot();
        });
    });

    // Common tests suite.
    commonTestsSuite(
        setup,
        { className: 'wrapper', prop: 'wrapper' },
        { className: CLASSNAME, prop: 'href', propValue: 'https://google.com' },
    );
});
