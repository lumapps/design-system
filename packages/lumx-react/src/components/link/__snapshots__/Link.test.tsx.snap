// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Link> Snapshots and structure should render a button if href attribute is missing 1`] = `
<button
  className="lumx-link"
  type="button"
/>
`;

exports[`<Link> Snapshots and structure should render color & color variant 1`] = `
<a
  className="lumx-link lumx-link--color-primary lumx-link--color-variant-D1"
  href="https://google.com"
/>
`;

exports[`<Link> Snapshots and structure should render correctly 1`] = `
<a
  className="lumx-link"
  href="https://google.com"
/>
`;

exports[`<Link> Snapshots and structure should render typography 1`] = `
<a
  className="lumx-link"
  href="https://google.com"
/>
`;
